R/gradethis_setup.R

```diff
@@ -0,0 +1,111 @@
+
+#' Setup gradethis for use within learnr
+#'
+#' Use this function to change the default options suggested by gradethis. This
+#' function also describes in detail each of the global options available for
+#' customization in the gradethis package. Note that you most likely do not want
+#' to change the defaults values for the learnr tutorial options that are
+#' prefixed with `exercise.`. Each of the gradethis-specific arguments sets a
+#' global option with the same name, prefixed with `gradethis.`. For example,
+#' `pass` sets `gradethis.pass`.
+#'
+#' @examples
+#' # Not run in package documentation because this function changes global opts
+#' if (FALSE) {
+#'   old_opts <- gradethis_setup(
+#'     pass = "Great work!", 
+#'     fail = "{random_encouragement()}"
+#'   )
+#' }
+#' 
+#' # Use getOption() to see the default value
+#' getOption("gradethis.pass")
+#' getOption("gradethis.code_correct")
+#' 
+#' @param pass Default message for [pass()]. Sets `options("gradethis.pass")`
+#' @param fail Default message for [fail()]. Sets `options("gradethis.fail")`
+#' @param code_correct Default `correct` message for [grade_this_code()]. If
+#'   unset, [grade_this_code()] falls back to the value of the `gradethis.pass`
+#'   option. Sets `options("gradethis.code_correct")`.
+#' @param code_incorrect Default `incorrect` message for [grade_this_code()]. If
+#'   unset [grade_this_code()] falls back to the value of the `gradethis.fail`
+#'   option. Sets `options("gradethis.code_incorrect")`.
+#' @param fail_code_feedback Logical `TRUE` or `FALSE` to determine whether 
+#'   [maybe_code_feedback()] should return code feedback, where if `FALSE`,
+#'   [maybe_code_feedback()] will return an empty string. This can be useful in
+#'   [pass()] or [fail()] messages and is used by default in the default 
+#'   [fail()] message and the default [grade_this_code()] incorrect message.
+#' @param allow_partial_matching Logical `TRUE` or `FALSE` to determine whether
+#'   partial matching is allowed in `grade_this_code()`. Sets
+#'   `options("gradethis.allow_partial_matching")`.
+#' @param pipe_warning The default message used in [pipe_warning()]. Sets
+#'   `options("gradethis.pipe_warning")`.
+#' @inheritParams learnr::tutorial_options
+#' @inheritDotParams learnr::tutorial_options
+#' 
+#' @return Invisibly returns the global options as they were prior to setting
+#'   them with `gradethis_setup()`.
+#' 
+#' @seealso [grade_learnr()]
+#' @export
+gradethis_setup <- function(
+  exercise.timelimit = NULL,
+  exercise.checker = NULL,
+  exercise.error.check.code = NULL,
+  ...,
+  pass = NULL,
+  fail = NULL,
+  code_correct = NULL,
+  code_incorrect = NULL,
+  fail_code_feedback = NULL,
+  pipe_warning = NULL,
+  allow_partial_matching = NULL
+) {
+  set_opts <- as.list(match.call()[-1])
```

Comment:
Maybe we also add a message if no arguments are provided at all, that they can remove the function call?

Reply:
Hmmm yeah I was thinking this is like elevator door close buttons, it shouldn't hurt to have it but it doesn't do anything if everything loaded properly.

Reply:
...but it makes people feel like they have a little more control :)

Reply:
Correct. Can leave as is as there is no harm in calling it with no parameters.  I guess if the examples are updated to remove calling `gradethis_setup()`, then I'm happy

Reply:
actually on second thought, maybe `gradethis_setup()` should _always_ set the default learnr options to make gradethis work. That way it always explicitly makes gradethis work, even if somehow the knitr options were messed up after `library(gradethis)` was called.

Reply:
or at least to always set the `exercise.checker` option